### PR TITLE
[test] wait(): Add TCP tests around wait()/close()

### LINKED ISSUE #577: [test] Integration Tests for TCP `wait()`
## Context

We should have the following integration tests for `wait()`:

- [ ] Attempt to wait on an invalid queue token.
- [x] Attempt to wait for an `accept()` operation to complete after closing a socket. b751b38d42f9decffa85a5031d9678168bdd5b5f
- [x] Attempt to wait for a `connect()` operation to complete after closing a socket. b751b38d42f9decffa85a5031d9678168bdd5b5f
- [x] Attempt to wait for an `accept()` operation to complete after asynchronously closing a socket. b751b38d42f9decffa85a5031d9678168bdd5b5f
- [x] Attempt to wait for a `connect()` operation to complete after asynchronously closing a socket. b751b38d42f9decffa85a5031d9678168bdd5b5f
- [ ] Attempt to wait for an `accept()` operation to complete after issuing an asynchronous close on a socket.
- [ ] Attempt to wait for a `connect()` operation to complete complete after asynchronous close on a socket.

### PR DESCRIPTION
**Pull request by anandbonde**
Closes #577

### CODE REVIEW

File: tests/rust/tcp-test/close/wait.rs [hunk 1]
```diff
@@ -224,3 +227,119 @@ fn wait_after_async_close_connecting_socket(libos: &mut LibOS, remote: &SocketAd
 
     Ok(())
 }
+
+fn wait_on_invalid_queue_token_returns_einval(libos: &mut LibOS) -> Result<()> {
```
Comment (ppenna) [inline tests/rust/tcp-test/close/wait.rs:231]: Could you add a one-line comment about this test here? You may want to copy what we have for description on the issue.

File: tests/rust/tcp-test/close/wait.rs [hunk 2]
```diff
@@ -224,3 +227,119 @@ fn wait_after_async_close_connecting_socket(libos: &mut LibOS, remote: &SocketAd
...
+    println!("{}", stringify!(wait_on_invalid_queue_token_returns_einval));
+
+    // Wait on an invalid queue token.
+    match libos.wait(QToken::from(u64::MAX), Some(Duration::from_micros(0))) {
```
Comment (ppenna) [inline tests/rust/tcp-test/close/wait.rs:235]: What about testing for some lower number, like zero, that is not a valid queue token as well?

File: tests/rust/tcp-test/close/wait.rs [hunk 3]
```diff
@@ -224,3 +227,119 @@ fn wait_after_async_close_connecting_socket(libos: &mut LibOS, remote: &SocketAd
...
+        Ok(_) => anyhow::bail!("wait() should not succeed on invalid token"),
+        Err(e) if e.errno == libc::EINVAL => {},
+        Err(_) => anyhow::bail!("wait() should not fail with any other reason than invalid token"),
+    }
+
+    Ok(())
+}
+
+fn wait_for_accept_after_issuing_async_close(libos: &mut LibOS, local: &SocketAddrV4) -> Result<()> {
```
Comment (ppenna) [inline tests/rust/tcp-test/close/wait.rs:244]: Could you add a one-line comment about this test here? You may want to copy what we have for description on the issue.

File: tests/rust/tcp-test/close/wait.rs [hunk 4]
```diff
@@ -224,3 +227,119 @@ fn wait_after_async_close_connecting_socket(libos: &mut LibOS, remote: &SocketAd
...
+        Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_ACCEPT && qr.qr_ret == 0 => {
+            anyhow::bail!("accept() should not succeed because remote should not be connecting")
+        },
+        Ok(_) => anyhow::bail!("wait() should not succeed with accept()"),
+        Err(_) => anyhow::bail!("wait() should timeout with accept()"),
+    }
+
+    let qt_close: QToken = libos.async_close(sockqd)?;
+
+    // Wait again on accept and ensure that ETIMEDOUT is returned.
```
Comment (ppenna) [inline tests/rust/tcp-test/close/wait.rs:266]: Should you reword `accept` to `accept()`?

File: tests/rust/tcp-test/close/wait.rs [hunk 5]
```diff
@@ -224,3 +227,119 @@ fn wait_after_async_close_connecting_socket(libos: &mut LibOS, remote: &SocketAd
...
+            anyhow::bail!("accept() should not succeed because remote should not be connecting")
+        },
+        Ok(_) => anyhow::bail!("wait() should return an error on accept() after close()"),
+        Err(e) => anyhow::bail!("wait() should not time out. {:?}", e),
+    }
+
+    Ok(())
+}
+
+fn wait_for_connect_after_issuing_async_close(libos: &mut LibOS, remote: &SocketAddrV4) -> Result<()> {
```
Comment (ppenna) [inline tests/rust/tcp-test/close/wait.rs:298]: Could you add a one-line comment about this test here? You may want to copy what we have for description on the issue.

File: tests/rust/tcp-test/close/wait.rs [hunk 6]
```diff
@@ -224,3 +227,119 @@ fn wait_after_async_close_connecting_socket(libos: &mut LibOS, remote: &SocketAd
...
+        }
+    }
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt_close, Some(Duration::from_micros(0))) {
+        Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_CLOSE && qr.qr_ret == 0 => {},
+        Ok(_) => anyhow::bail!("wait() should succeed with async_close()"),
+        Err(e) => anyhow::bail!("wait() should succeed. {:?}", e),
+    }
+
```
Comment (ppenna) [inline tests/rust/tcp-test/close/wait.rs:343]: Don't we want to poll again the `connect()` queue token to ensure that it is cancelled?

**Review by ppenna — APPROVED** [review]
LGTM.

File: tests/rust/tcp-test/close/wait.rs (hunk 3 above)
Comment (anandbonde) [reply tests/rust/tcp-test/close/wait.rs:253]: Done.

File: tests/rust/tcp-test/close/wait.rs (hunk 4 above)
Comment (anandbonde) [reply tests/rust/tcp-test/close/wait.rs]: Done.

File: tests/rust/tcp-test/close/wait.rs (hunk 5 above)
Comment (anandbonde) [reply tests/rust/tcp-test/close/wait.rs:308]: Done.

**Pull request merged**